stock_storage_type_putaway_strategy/models/stock_location.py

```diff
@@ -0,0 +1,199 @@
+# Copyright 2019 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+from odoo.osv.expression import AND, OR
+
+
+class StockLocation(models.Model):
+
+    _inherit = "stock.location"
+
+    pack_putaway_strategy = fields.Selection(
+        selection=[
+            ("none", "None"),
+            ("ordered_locations", "Ordered Children Locations"),
+        ],
+        required=True,
+        default="none",
+        string="Packs Put-Away Strategy",
+        help="This defines the storage strategy to use when packs are put "
+        "away in this location.\n"
+        "None: when a pack is moved to this location, it will not be put"
+        " away any further.\n"
+        "Ordered Children Locations: when a pack is moved to this "
+        "location, a suitable location will be searched in its children "
+        "locations according to the restrictions defined on their "
+        "respective location storage types.",
+    )
+    storage_location_sequence_ids = fields.One2many(
+        "stock.storage.location.sequence",
+        "location_id",
+        string="Storage locations sequences",
+    )
+
+    @api.constrains("pack_putaway_strategy", "storage_location_sequence_ids")
+    def _check_pack_putaway_strategy(self):
+        for location in self:
+            if (
+                location.pack_putaway_strategy == "none"
+                and location.storage_location_sequence_ids
+            ):
+                raise ValidationError(
+                    _(
+                        "Changing Packs storage strategy to 'None' is not "
+                        "allowed as the location %s is used in a Storage "
+                        "locations for package storage type."
+                    )
+                    % location.name
+                )
+
+    def _get_putaway_strategy(self, product):
+        putaway_location = super()._get_putaway_strategy(product)
+        quant = self.env.context.get("storage_quant")
+        return self._get_pack_putaway_strategy(
+            putaway_location, quant, product
+        )
+
+    def _get_pack_putaway_strategy(self, putaway_location, quant, product):
+        package_storage_type = False
+        if quant:
+            package_storage_type = quant.package_id.package_storage_type_id
```

Comment:
if package_storage_type_id exists on package_id take it, otherwise look at the one stored on the quant